syntax = "proto3";

package order;

import "product.proto";

service OrderService {
  rpc CreateOrder(CreateOrderRequest) returns (OrderResponse);
  rpc GetOrder(OrderRequest) returns (OrderResponse);
  rpc GetUserOrders(UserOrdersRequest) returns (UserOrdersResponse);
  rpc UpdateOrderStatus(UpdateOrderStatusRequest) returns (OrderResponse);
  rpc CancelOrder(CancelOrderRequest) returns (OrderResponse);
}

message OrderItem {
  string product_id = 1;
  int32 quantity = 2;
  double price = 3;
}

message CreateOrderRequest {
  string user_id = 1;
  repeated OrderItem items = 2;
  string shipping_address = 3;
  string payment_method = 4;
}

message OrderRequest {
  string order_id = 1;
}

message UserOrdersRequest {
  string user_id = 1;
  int32 page = 2;
  int32 limit = 3;
}

message UpdateOrderStatusRequest {
  string order_id = 1;
  string status = 2;
  string admin_id = 3;
}

message CancelOrderRequest {
  string order_id = 1;
  string user_id = 2;
}

message OrderResponse {
  string order_id = 1;
  string user_id = 2;
  repeated OrderItem items = 3;
  double total_amount = 4;
  string status = 5;
  string shipping_address = 6;
  string payment_method = 7;
  string created_at = 8;
  string updated_at = 9;
  string message = 10;
}

message UserOrdersResponse {
  repeated OrderResponse orders = 1;
  int32 total = 2;
  int32 page = 3;
  int32 limit = 4;
}
